new-components/scanners/gosec/internal/transformer/transformer.go

```diff
@@ -0,0 +1,391 @@
+package transformer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/smithy-security/pkg/env"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	schemav1 "github.com/smithy-security/pkg/sarif/spec/gen/sarif-schema/v2-1-0"
+
+	"github.com/smithy-security/smithy/sdk/component"
+	ocsf "github.com/smithy-security/smithy/sdk/gen/com/github/ocsf/ocsf_schema/v1"
+	ocsffindinginfo "github.com/smithy-security/smithy/sdk/gen/ocsf_ext/finding_info/v1"
+
+	"github.com/smithy-security/smithy/new-components/scanners/gosec/internal/util/ptr"
+)
+
+const TargetTypeRepository TargetType = "repository"
+
+type (
+	// GosecTransformerOption allows customising the transformer.
+	GosecTransformerOption func(g *gosecTransformer) error
+
+	// TargetType represents the target type.
+	TargetType string
+
+	gosecTransformer struct {
+		targetType     TargetType
+		clock          clockwork.Clock
+		rawOutFilePath string
+	}
+)
+
+func (tt TargetType) String() string {
+	return string(tt)
+}
+
+// GosecTransformerWithClock allows customising the underlying clock.
+func GosecTransformerWithClock(clock clockwork.Clock) GosecTransformerOption {
+	return func(g *gosecTransformer) error {
+		if clock == nil {
```

Comment:
what is have in mind is this:
```golang

var ErrInvalidClock = errors.New("invalid nil clock")

func GosecTransformerWithClock(clock clockwork.Clock) (GosecTransformerOption, error) {
	if clock == nil {
		return nil, ErrInvalidClock
	}
	return func(g *gosecTransformer) error {
		g.clock = clock
		return nil
	}
}
```
This way if the clock is nil, the function will fail fast instead of failing when it's invoked inside the transformer constructor

Reply:
I think that calling this before the designed loop for function would add more code for little gains

This pattern works quite well, is well known and follows the standard that I made in the changes

Some sources:
- https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis
- https://michalzalecki.com/golang-options-pattern/

If you don't mind I'll keep this as is :)